File: reference-implementation/go/account.go
package main

import (
	"context"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

func registerAccountTools(s *server.MCPServer) {
	registerAccountToolsWithState(s, state)
}

func registerAccountToolsWithState(s *server.MCPServer, st *referenceState) {
	s.AddTool(
		mcp.NewTool("apex.account.summary",
			mcp.WithDescription("Current account state — balances, margin utilisation, equity."),
			mcp.WithString("account_id", mcp.Required()),
			mcp.WithString("currency", mcp.Description("Response currency. Defaults to account base currency.")),
			mcp.WithReadOnlyHintAnnotation(true),
			mcp.WithDestructiveHintAnnotation(false),
			mcp.WithIdempotentHintAnnotation(true),
		),
		func(_ context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			args := request.GetArguments()
			accountID := strParam(args, "account_id", "")
			if accountID == "" {
				return jsonResult(apexError("APEX_4011", "validation", "account_id is required"))
			}
			return jsonResult(st.accountSummary(strParam(args, "currency", "USD")))
		},
	)

	s.AddTool(
		mcp.NewTool("apex.account.positions",
			mcp.WithDescription("All open positions with live P&L."),
			mcp.WithString("account_id", mcp.Required()),
			mcp.WithString("instrument_id", mcp.Description("APEX canonical instrument ID (e.g. APEX:FX:EURUSD)")),
			mcp.WithString("profile", mcp.Description("Asset class profile filter")),
			mcp.WithReadOnlyHintAnnotation(true),
			mcp.WithDestructiveHintAnnotation(false),
			mcp.WithIdempotentHintAnnotation(true),
		),
		func(_ context.Context, _ mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			return jsonResult(st.positionsResponse())
		},
	)

	s.AddTool(
		mcp.NewTool("apex.account.orders",
			mcp.WithDescription("Known orders and their current lifecycle state."),
			mcp.WithString("account_id", mcp.Required()),
			mcp.WithString("status", mcp.Description("Filter by order status"), mcp.Enum("working", "partially_filled", "filled", "cancelled", "rejected", "expired", "all")),
			mcp.WithString("instrument_id", mcp.Description("APEX canonical instrument ID (e.g. APEX:FX:EURUSD)")),
			mcp.WithReadOnlyHintAnnotation(true),
			mcp.WithDestructiveHintAnnotation(false),
			mcp.WithIdempotentHintAnnotation(true),
		),
		func(_ context.Context, _ mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			return jsonResult(st.ordersResponse())
		},
	)

	s.AddTool(
		mcp.NewTool("apex.account.history",
			mcp.WithDescription("Closed trades and funding events."),
			mcp.WithString("account_id", mcp.Required()),
			mcp.WithString("from", mcp.Required(), mcp.Description("ISO8601 start date")),
			mcp.WithString("to", mcp.Required(), mcp.Description("ISO8601 end date")),
			mcp.WithString("event_type", mcp.Description("Event type filter"), mcp.Enum("trade", "funding", "cash", "corporate_action", "all")),
			mcp.WithNumber("limit", mcp.Description("Maximum number of events to return (1-500, default 100)")),
			mcp.WithString("cursor", mcp.Description("Pagination cursor")),
			mcp.WithReadOnlyHintAnnotation(true),
			mcp.WithDestructiveHintAnnotation(false),
			mcp.WithIdempotentHintAnnotation(true),
		),
		handleAccountHistory,
	)
}

func handleAccountHistory(_ context.Context, _ mcp.CallToolRequest) (*mcp.CallToolResult, error) {
	return jsonResult(historyResponse{
		Events:     []any{},
		NextCursor: nil,
		HasMore:    false,
	})
}

File: reference-implementation/typescript/src/tools/account.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

import { nowIso } from "../lib/helpers.js";
import type { ReferenceTradingState } from "../lib/resources.js";
import { InstrumentIdSchema } from "../lib/schemas.js";

export function registerAccountTools(server: McpServer, state: ReferenceTradingState): void {
  server.registerTool(
    "apex.account.summary",
    {
      description: "Current account state — balances, margin utilisation, equity.",
      inputSchema: {
        account_id: z.string(),
        currency: z.string().optional().describe("Response currency. Defaults to account base currency."),
      },
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    },
    async ({ account_id, currency }) => {
      const summary = state.getAccountSummary();
      return {
        structuredContent: {
          ...summary,
          account_id,
          response_currency: currency ?? "USD",
        },
        content: [],
      };
    },
  );

  server.registerTool(
    "apex.account.positions",
    {
      description: "All open positions with live P&L.",
      inputSchema: {
        account_id: z.string(),
        instrument_id: InstrumentIdSchema.optional(),
        profile: z.string().optional(),
      },
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    },
    async ({ instrument_id }) => {
      const positions = state.getPositions();
      return {
        structuredContent: {
          ...positions,
          positions: positions.positions.map((position) => ({
            ...position,
            instrument_id: instrument_id ?? position.instrument_id,
          })),
        },
        content: [],
      };
    },
  );

  server.registerTool(
    "apex.account.orders",
    {
      description: "Known orders and their current lifecycle state.",
      inputSchema: {
        account_id: z.string(),
        status: z
          .enum(["working", "partially_filled", "filled", "cancelled", "rejected", "expired", "all"])
          .default("all"),
        instrument_id: InstrumentIdSchema.optional(),
      },
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    },
    async () => ({
      structuredContent: state.getOrders(),
      content: [],
    }),
  );

  server.registerTool(
    "apex.account.history",
    {
      description: "Closed trades and funding events.",
      inputSchema: {
        account_id: z.string(),
        from: z.string().describe("ISO8601 start date"),
        to: z.string().describe("ISO8601 end date"),
        event_type: z.enum(["trade", "funding", "cash", "corporate_action", "all"]).default("all"),
        limit: z.number().int().min(1).max(500).default(100),
        cursor: z.string().optional().describe("Pagination cursor"),
      },
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    },
    async () => ({
      structuredContent: { events: [], next_cursor: null, has_more: false },
      content: [],
    }),
  );
}
